howso/feature_residuals.amlg

```diff
@@ -1102,6 +1191,152 @@
+						removed_features
+							(filter
+								(lambda
+									(not (contains_value
+										(indices compressed_values)
+										(current_value)
+									))
+								)
+								(indices !sharedDeviationsMap)
+							)
+					)
+					(assign (assoc
+						compressed_values
+							(append
+								compressed_values
+								;grab the values for the removed features by looking up their primary key values
+								(map
+									(lambda
+										(get compressed_values (get !sharedDeviationsMap (current_index)))
+									)
+									(zip removed_features)
+								)
+							)
+					))
+				)
+			)
+		)
+		compressed_values
+	)
+
+	;Helper function to prep shared feature deviations, by taking in a list of features and their list of residuals,
+	; combining all of the residuals under the primary feature key of the group of the shared deviations, and removing the features
+	; from the features list whose residuals have been combined into another features.
+	;parameters:
+	; features_list: A features list
+	; feature_residuals_lists: A features residuals whose lists are in the order of the features_list
+	#!PrepSharedDeviations
+	(declare
+		(assoc
+			features_list (assoc)
+			feature_residuals_lists (assoc)
+		)
+
+		(declare (assoc
+			joined_original_features_residuals_list (zip features_list feature_residuals_lists)
+		))
+
+		;assoc of features and their residuals that are not in a shared deviations group
+		(declare (assoc
+			non_shared_deviations_features_list
+				(remove joined_original_features_residuals_list (indices !sharedDeviationsMap))
+		))
+
+		;assoc of primary shared deviations features and their residuals that are combined from all the other features in their shared deviations group
+		(declare (assoc
+			shared_deviations_residuals_list_combined
+				(map
+					(lambda
+						(let
+							(assoc
+								;the group of features in the shared deviations group
+								feature_group (call !GetSharedDeviationGrouping (assoc feature_group_to_retrieve (current_index 2)))
+								current_feature (current_index 1)
+							)
+							(apply "append"
+								(map
+									(lambda
+										(let
+											(assoc
+												current_feature_residuals_list
+													(get joined_original_features_residuals_list (current_value 1))
+											)
+											(if (= (null) current_feature_residuals_list)
+												[]
+												current_feature_residuals_list
+											)
+										)
+									)
+									feature_group
+								)
+							)
+						)
+					)
+					;group primary features
+					(zip (values !sharedDeviationsMap))
+				)
+		))
+
+		(assign (assoc
+			shared_deviations_residuals_list_combined
+				;the combined list has a list for all feature deviations groups, this filters out groups with no values
+				(filter
+					(lambda (!= [] (current_value)))
+					shared_deviations_residuals_list_combined
+				)
+		))
+
+		;add the two assoc back together
+		(declare (assoc
```

Comment:
you could create this mapping by taking `joined_original_features_residuals_list` (or whatever you rename it to), removing all indices from it that are !sharedDeviationsMap and then append `shared_deviations_residuals_list_combined` (or whatever you rename it to), that way you completely get rid of `non_shared_deviations_features_list`